Source/SwiftyDropbox/Shared/Handwritten/OAuth/AuthSession.swift

```diff
@@ -0,0 +1,104 @@
+    ///       To call one API endpoint you have to obtains the scope first otherwise you will get HTTP 401.
+    ///     - includeGrantedScopes: If false, Dropbox will give you the scopes in scopes array.
+    ///       Otherwise Dropbox server will return a token with all scopes user previously granted your app
+    ///       together with the new scopes.
+    public init(scopeType: ScopeType, scopes: [String], includeGrantedScopes: Bool) {
+        self.scopeType = scopeType
+        self.scopes = scopes
+        self.includeGrantedScopes = includeGrantedScopes
+    }
+}
+
+// MARK: Internal
+
+/// Object that contains all the necessary data of an OAuth 2 Authorization Code Flow with PKCE.s
+struct AuthSession {
+    let scopeRequest: ScopeRequest?
+    let pkceData: PkceData
+    let state: String
+    let tokenAccessType = "offline"
+    let responseType = "code"
+
+    init(scopeRequest: ScopeRequest?) {
+        self.pkceData = PkceData()
+        self.scopeRequest = scopeRequest
+        self.state = Self.createState(with: pkceData, scopeRequest: scopeRequest, tokenAccessType: tokenAccessType)
+    }
+
+    private static func createState(
+        with pkceData: PkceData, scopeRequest: ScopeRequest?, tokenAccessType: String
+    ) -> String {
+        let codeChallenge = pkceData.codeChallenge
+        var state = "oauth2code:\(codeChallenge):\(pkceData.codeChallengeMethod):\(tokenAccessType)"
+        if let scopeRequest = scopeRequest {
+            if let scopeString = scopeRequest.scopeString {
+                state += ":\(scopeString)"
+            }
+            if scopeRequest.includeGrantedScopes {
+                state += ":\(scopeRequest.scopeType.rawValue)"
+            }
+        }
+        return state
+    }
+}
+
+/// PKCE data for OAuth 2 Authorization Code Flow.
+struct PkceData {
+    let codeVerifier: String
+    let codeChallenge: String
+    let codeChallengeMethod = "S256"
+
+    init() {
+        self.codeVerifier = Self.randomStringOfLength(128)
+        self.codeChallenge = Self.codeChallengeFromCodeVerifier(codeVerifier)
+    }
+
+    private static func randomStringOfLength(_ length: Int) -> String {
+        let alphanumerics = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+        return String((0..<length).map { _ in alphanumerics.randomElement()! })
+    }
+
+    private static func codeChallengeFromCodeVerifier(_ codeVerifier: String) -> String {
+        let data = codeVerifier.data(using: .ascii)!
+        var digest = [UInt8](repeating: 0, count:Int(CC_SHA256_DIGEST_LENGTH))
+        _ = data.withUnsafeBytes {
+            CC_SHA256($0.baseAddress, UInt32(data.count), &digest)
+        }
+        /// Replace these characters to make the string safe to use in a URL.
+        return Data(bytes: digest).base64EncodedString()
+            .replacingOccurrences(of: "/", with: "_")
+            .replacingOccurrences(of: "+", with: "-")
+            .replacingOccurrences(of: "=", with: "")
```

Comment:
https://developer.apple.com/documentation/foundation/nsstring/1411946-stringbyaddingpercentencodingwit

with https://developer.apple.com/documentation/foundation/nscharacterset/1416698-urlqueryallowedcharacterset?language=objc

Reply:
I meant to make a base64 string url safe, not to add percent encoding.
spec here https://tools.ietf.org/html/rfc4648#page-7.